<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta charset="utf-8">
    <title>DigenicSearch</title>

    <!-- Latest compiled and minified CSS -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">

    <!-- Optional theme -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">

    <!-- jQuery (necessary for Bootstrap's JavaScript plugins) -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>

    <!-- Latest compiled and minified JavaScript -->
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link href="css/docs.css" rel="stylesheet">

  </head>

  <body>
  <nav class="navbar navbar-default navbar-fixed-top">
    <div class="container-fluid">
      <div class="navbar-header">
        <div class="navbar-header">
          <a class="navbar-brand" href="#">DigenicSearch</a>
          <ul class="nav navbar-nav navbar-right">
            <li class="navbar-text" id="Version">
              Version</li>
          </ul>
        </div>
      </div>



    </div>

  </nav>
  <div class="container">

    <div class="row">
      <div class="col-sm-3 col-md-2 sidebar" >

        <ul class="nav nav-sidebar nav-list">
          <li><b>Index</b></li>
          
          <li><a href="#description"><span>Description</span></a></li>
          
          <li><a href="#installation"><span>Installation</span></a></li>
          
          <li><a href="#manual"><span>Manual</span></a></li>
          
          <li><a href="#about"><span>About</span></a></li>
          
          <li><a href="#contributing"><span>Contributing</span></a></li>
          
          <li><a href="#contact"><span>Contact</span></a></li>
          
        </ul>
        <ul class="nav nav-sidebar nav-list">
          <li><b>Project Links</b></li>

          <li><a href="https://github.com/biopet/digenicsearch">Source code</a></li>
          <li><a href="https://github.com/biopet/digenicsearch/issues">Feedback, feature requests and bug reports</a></li>
          <li><a href="api/index.html">API Documentation</a></li>
        </ul>
        <ul class="nav nav-sidebar nav-list">
          <li><b>Documentation</b></li>
          <li><a href="../index.html">Latest stable version</a></li>
          <li><a href="../develop/index.html">Development version</a></li>
        </ul>
      </div>
      
      <div class="col-sm-9 col-sm-offset-3 col-md-10 col-md-offset-2 main" id="top">

        
        <h1 id="description" class="section">Description</h1>
        <p>This will will search for a combination of variants within a multi sample vcf file.
        The tool can filter on INFO fields and a maximum distance of the snps on the reference.</p>
        
        <h1 id="installation" class="section">Installation</h1>
        <p>DigenicSearch requires Java 8 to be installed on your device. Download Java 8
        <a href="http://www.oracle.com/technetwork/java/javase/downloads/jre8-downloads-2133155.html">here</a>
        or install via your distribution&#39;s package manager.</p>
        <p>Download the latest version of DigenicSearch <a href="https://github.com/biopet/digenicsearch/releases/">here</a>.
        To generate the usage run:</p>
        <pre><code>java -jar &lt;DigenicSearch_jar&gt; --help</code></pre>
        
        <h1 id="manual" class="section">Manual</h1>
        <p>Because of the number of possible combination this tool requires to run on a spark cluster.
        If required the tool can still run local by submitting the tool to a local master, see also https://spark.apache.org/docs/latest/submitting-applications.html#master-urls
        By default this tool runs on the complete genome but with the option --regions a bed file can be provided to limit the number of locations</p>
        
        <h2 id="example" class="section">Example</h2>
        <p>A default run:</p>
        <pre><code>java -jar &lt;DigenicSearch_jar&gt; \
-i &lt;input vcf&gt; \
-o &lt;output dir&gt; \
-R &lt;reference fasta&gt; \
-p &lt;ped file</code></pre>
        <p>A run on limited locations:</p>
        <pre><code>java -jar &lt;DigenicSearch_jar&gt; \
-i &lt;input vcf&gt; \
-o &lt;output dir&gt; \
-R &lt;reference fasta&gt; \
--regions &lt;bed file&gt; \
-p &lt;ped file</code></pre>
        
        <h2 id="usage" class="section">Usage</h2>
        <p>Usage for DigenicSearch:</p>
        <table class="table">
          <thead>
            <tr>
              <th>Option</th>
              <th>Required</th>
              <th>Can occur multiple times</th>
              <th>Description</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>--log_level, -l</td>
              <td>no</td>
              <td>no</td>
              <td>Level of log information printed. Possible levels: &#39;debug&#39;, &#39;info&#39;, &#39;warn&#39;, &#39;error&#39;</td>
            </tr>
            <tr>
              <td>--help, -h</td>
              <td>no</td>
              <td>no</td>
              <td>Print usage</td>
            </tr>
            <tr>
              <td>--version, -v</td>
              <td>no</td>
              <td>no</td>
              <td>Print version</td>
            </tr>
            <tr>
              <td>--inputFile, -i</td>
              <td>yes</td>
              <td>no</td>
              <td>Input vcf files</td>
            </tr>
            <tr>
              <td>--outputDir, -o</td>
              <td>yes</td>
              <td>no</td>
              <td>Output dir for the tool</td>
            </tr>
            <tr>
              <td>--reference, -R</td>
              <td>yes</td>
              <td>no</td>
              <td>Reference fasta file to use, dict file should be next to it</td>
            </tr>
            <tr>
              <td>--regions</td>
              <td>no</td>
              <td>no</td>
              <td>Only using this regions in the bed file</td>
            </tr>
            <tr>
              <td>--pedFile, -p</td>
              <td>yes</td>
              <td>yes (unlimited)</td>
              <td>Input ped file for family relations and effected/non-effected</td>
            </tr>
            <tr>
              <td>--detectionMode</td>
              <td>no</td>
              <td>no</td>
              <td>Detection mode, possible values: Varant, Allele, Genotype</td>
            </tr>
            <tr>
              <td>--singleAnnotationFilter</td>
              <td>no</td>
              <td>no</td>
              <td>Filter on single variant</td>
            </tr>
            <tr>
              <td>--pairAnnotationFilter</td>
              <td>no</td>
              <td>no</td>
              <td>Filter on paired variant, must be true for 1 of the 2 in the pair</td>
            </tr>
            <tr>
              <td>--singleAffectedFraction</td>
              <td>no</td>
              <td>no</td>
              <td>minimal affected fraction for each variant</td>
            </tr>
            <tr>
              <td>--pairAffectedFraction</td>
              <td>no</td>
              <td>no</td>
              <td>minimal affected fraction for for at least 1 of the 2 variants</td>
            </tr>
            <tr>
              <td>--singleUnaffectedFraction</td>
              <td>no</td>
              <td>no</td>
              <td>maximum unaffected fraction for for each variant</td>
            </tr>
            <tr>
              <td>--pairUnaffectedFraction</td>
              <td>no</td>
              <td>no</td>
              <td>maximum unaffected fraction for for at least 1 of the 2 variants</td>
            </tr>
            <tr>
              <td>--maxDistance</td>
              <td>no</td>
              <td>no</td>
              <td>maxDistance in base pairs. This option will make the assumption that both variants are on the same contig</td>
            </tr>
            <tr>
              <td>--binSize</td>
              <td>no</td>
              <td>no</td>
              <td>Binsize in estimated base pairs</td>
            </tr>
            <tr>
              <td>--maxContigsInSingleJob</td>
              <td>no</td>
              <td>no</td>
              <td>Max number of bins to be combined, default is 250</td>
            </tr>
            <tr>
              <td>--sparkMaster</td>
              <td>no</td>
              <td>no</td>
              <td>Spark master, default to local[1]</td>
            </tr>
          </tbody>
        </table>
        
        <h1 id="about" class="section">About</h1>
        <p>DigenicSearch is part of BIOPET tool suite that is developed at LUMC by <a href="http://sasc.lumc.nl/">the SASC team</a>.
        Each tool in the BIOPET tool suite is meant to offer a standalone function that can be used to perform a
        dedicate data analysis task or added as part of <a href="http://biopet-docs.readthedocs.io/en/latest/">BIOPET pipelines</a>.</p>
        <p>All tools in the BIOPET tool suite are <a href="https://www.gnu.org/philosophy/free-sw.html">Free/Libre</a> and
        <a href="https://opensource.org/osd">Open Source</a> Software.</p>
        
        <h1 id="contributing" class="section">Contributing</h1>
        <p>The source code of DigenicSearch can be found <a href="https://github.com/biopet/digenicsearch">here</a>.
        We welcome any contributions. Bug reports, feature requests and feedback can be submitted at our
        <a href="https://github.com/biopet/digenicsearch/issues">issue tracker</a>.</p>
        <p>DigenicSearch is build using sbt. Before submitting a pull request, make sure all tests can be passed by
        running <code>sbt test</code> from the project&#39;s root. We recommend using an IDE to work on DigenicSearch. We have had
        good results with <a href="https://www.jetbrains.com/idea/">this IDE</a>.</p>
        
        <h1 id="contact" class="section">Contact</h1>
        <p>
          <!-- Obscure e-mail address for spammers -->
        For any question related to DigenicSearch, please use the
        <a href='https://github.com/biopet/digenicsearch/issues'>github issue tracker</a>
        or contact
         <a href='http://sasc.lumc.nl/'>the SASC team</a> directly at: <a href='&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#115;&#97;&#115;&#99;&#64;&#108;&#117;&#109;&#99;&#46;&#110;&#108;'>
        &#115;&#97;&#115;&#99;&#64;&#108;&#117;&#109;&#99;&#46;&#110;&#108;</a>.
        </p>      

      </div>
    </div>

  </div>
  <script>
    var loc = window.location.pathname;
    var thisDir = loc.substring(0, loc.lastIndexOf('/'));
    var splitLocation = window.location.pathname.split('/');
    var version = splitLocation[splitLocation.length - 2]
    document.getElementById("Version").innerHTML = "Version " + version;
  </script>
</body>

</html>
